### PR TITLE
[CSL-610] Add test for VssCertificate rollback

### PR DESCRIPTION
**Pull request by rockbmb**

### CODE REVIEW

File: src/Pos/Ssc/GodTossing/Core/Types.hs [hunk 1]
```diff
@@ -162,7 +166,10 @@ instance Ord VssCertificate where
 
 instance Buildable VssCertificate where
     build VssCertificate {..} = bprint
-        ("vssCert:"%build%":"%int) vcSigningKey vcExpiryEpoch
+        ("vssCert:"%build%"   ;;ee:"%int%" key:"%build%"sig:"%build) vcSigningKey
```
Comment (rockbmb) [inline src/Pos/Ssc/GodTossing/Core/Types.hs]: Alright, this is a leftover test marker. Will remove and force push.

File: src/Pos/Ssc/GodTossing/Toss/Pure.hs [hunk 2]
```diff
@@ -63,6 +67,21 @@ instance MonadToss PureToss where
...
+runPureToss
+    :: LoggerName
+    -> MultiRichmenStake
+    -> GtGlobalState
+    -> PureToss a
+    -> (a, GtGlobalState, [LogEvent])
+runPureToss loggerName richmenData gs =
+    convertRes .
+    (\a -> runRWS a richmenData gs) .
+    usingLoggerName loggerName . runPureLog . runNamedPureLogger . getPureToss
```
Comment (ChShersh) [inline src/Pos/Ssc/GodTossing/Toss/Pure.hs]: There're some utility functions that helps reduce code amount:
https://github.com/serokell/log-warper/blob/c2a21fd2971d9ed48cdd7295f8efd56fc8ec4821/src/System/Wlog/CanLog.hs#L155

File: test/Test/Pos/Ssc/GodTossing/VssCertDataSpec.hs [hunk 3]
```diff
@@ -34,6 +40,29 @@ spec = describe "Ssc.GodTossing.VssCertData" $ do
         prop description_verifySetLastKnownSlot verifySetLastKnownSlot
     describe "verifyDeleteAndFilter" $
         prop description_verifyDeleteAndFilter verifyDeleteAndFilter
+    let smaller = modifyMaxSuccess (const 100)
+    describe "verifyRollback" $ smaller $
+        prop description_verifyRollback $
+            do goodVssCertData@(VssCertData {..}) <- getVssCertData <$> arbitrary
```
Comment (ChShersh) [inline test/Test/Pos/Ssc/GodTossing/VssCertDataSpec.hs]: I think it's better to move this code into separate function as it's done usually.

File: src/Pos/Ssc/GodTossing/Toss/Pure.hs (hunk 2 above)
Comment (rockbmb) [inline src/Pos/Ssc/GodTossing/Toss/Pure.hs]: This appears to require `instance (HasLoggerName Identity)`. Should I add this in `log-warper` or in here in this module? Furthermore, what should the identity monad's logger name be?

Comment (ChShersh) [inline src/Pos/Ssc/GodTossing/Toss/Pure.hs]: @rockbmb I think you can easily add such instance directly into `log-warper`.

File: src/Pos/Ssc/GodTossing/Toss/Pure.hs [hunk 4]
```diff
@@ -37,10 +35,10 @@ newtype PureToss a = PureToss
     { getPureToss :: NamedPureLogger (RWS MultiRichmenStake () GtGlobalState) a
     } deriving (Functor, Applicative, Monad, CanLog, HasLoggerName)
 
-instance (Monad m, HasLoggerName m) => HasLoggerName (IdentityT a m) where
-    getLoggerName = lift getLoggerName
+instance HasLoggerName Identity where
```
Comment (rockbmb) [inline src/Pos/Ssc/GodTossing/Toss/Pure.hs]: @ChShersh I'll move this to `log-warper` if I also end up needing it in `PurePoll` for CSL-979.

**Comment by rockbmb** [comment]
@ChShersh I merged `master` to trigger CI. If you're not waiting for someone else's review, can you merge and close both this and the CSL-843 PR?

**Comment by ChShersh** [comment]
@rockbmb Merged

File: src/Pos/Ssc/GodTossing/Toss/Class.hs [hunk 5]
```diff
@@ -57,6 +61,10 @@ class (Monad m, WithLogger m) =>
         m SharesMap
     getShares = lift getShares
 
+    default getVssCertData :: (MonadTrans t, MonadTossRead m', t m' ~ m) =>
+        m VssCertData
+    getVssCertData = lift getVssCertData
+
     default getVssCertificates :: (MonadTrans t, MonadTossRead m', t m' ~ m) =>
         m VssCertificatesMap
     getVssCertificates = lift getVssCertificates
```
Comment (pva701) [inline src/Pos/Ssc/GodTossing/Toss/Class.hs]: @rockbmb you could make default definition: `getVssCertificates = certs <$> getVssCertData`

File: src/Pos/Ssc/GodTossing/Toss/Trans.hs [hunk 6]
```diff
@@ -89,8 +90,8 @@ instance MonadTossRead m =>
     getCommitments = TossT $ (<>) <$> use tmCommitments <*> getCommitments
     getOpenings = TossT $ (<>) <$> use tmOpenings <*> getOpenings
     getShares = TossT $ (<>) <$> use tmShares <*> getShares
-    getVssCertificates =
-        TossT $ (<>) <$> use tmCertificates <*> lift getVssCertificates
+    getVssCertificates = certs <$> getVssCertData
```
Comment (gromakovsky) [inline src/Pos/Ssc/GodTossing/Toss/Trans.hs]: It was wrong change. You should consider `tmCertificates` in `TossT`. It was fixed in #1160.

**Pull request merged**